subsys/bluetooth/audio/has.c

```diff
@@ -0,0 +1,1330 @@
+/*
+ * Copyright (c) 2022 Codecoup
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+
+#include <device.h>
+#include <errno.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <string.h>
+#include <net/buf.h>
+#include <sys/atomic.h>
+#include <sys/byteorder.h>
+#include <sys/check.h>
+#include <sys/math_extras.h>
+
+#include <bluetooth/audio/has.h>
+#include <bluetooth/bluetooth.h>
+#include <bluetooth/conn.h>
+#include <bluetooth/gatt.h>
+
+#include "has_internal.h"
+#include "../host/conn_internal.h"
+#include "../host/hci_core.h"
+
+#define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_DEBUG_HAS)
+#define LOG_MODULE_NAME bt_has
+#include "common/log.h"
+
+#define BT_HAS_MAX_CLIENT MIN(CONFIG_BT_MAX_CONN, CONFIG_BT_MAX_PAIRED)
+#define BT_HAS_CP_WORK_TIMEOUT K_MSEC(10)
+#define BT_HAS_CP_TX_MAX_LEN                                                                       \
+	(1 + 2 /* Preset Changed */ + 3 /* Generic Update */ + BT_HAS_PRESET_NAME_MAX)
+
+enum {
+	CLIENT_FLAG_ENCRYPTED, /* if ACL is encrypted */
+	CLIENT_FLAG_ATT_MTU_VALID, /* if ATT MTU meet HAS specification requirements */
+	CLIENT_FLAG_CP_IND_ENABLED, /* if Control Point indications are enabled */
+	CLIENT_FLAG_CP_NFY_ENABLED, /* if Control Point notifications are enabled */
+	CLIENT_FLAG_CP_BUSY, /* if Control Point operation is in progress */
+
+	/* Total number of flags - must be at the end */
+	CLIENT_NUM_FLAGS,
+};
+
+#if defined(CONFIG_BT_HAS)
+static struct has_preset {
+	uint8_t id;
```

Comment:
I am confused about the relation between "id" and "index".
Here, the preset has an "id", which, from elsewhere in the code, seems to be the same as the "index" of a preset.
Ditto for the `last_preset_id` in line 97.
But "index" also seems to be used for the same thing, e.g. in the `preset_foreach` function.  And in some calls to that function, "id"s are supplied where "index"es are specified as parameters.
It seems to me that these are the same, and then they should have the same name.  Preferably "index", as that is what is used in the spec.  (Although I think that is poor naming in the spec.)

To add to the confusion, there is also the "client_index", which seems to be an actual index used to index the array of clients.